File: tests/test_problems.cpp
#include <catch2/catch_test_macros.hpp>
#include <catch2/matchers/catch_matchers.hpp>
#include <catch2/matchers/catch_matchers_exception.hpp>
#include <catch2/matchers/catch_matchers_string.hpp>

#include <nlohmann/json.hpp>

#include <fstream>
#include <sstream>

#include "helpers/fake_transport.hpp"
#include "helpers/fixtures.hpp"
#include "lcpush/errors.hpp"
#include "lcpush/paths.hpp"
#include "lcpush/problems.hpp"
#include "lcpush/util/time.hpp"

using namespace lcpush;
using lcpush::testing::IsolatedDirs;
using lcpush::testing::json_response;
using nlohmann::json;

namespace {

// 2026-07-27T10:00:00Z, matching the Python test's NOW.
const std::time_t kNow = *util::parse_iso_utc("2026-07-27T10:00:00Z");
constexpr std::time_t kDay = 24 * 3600;

json page(int total, int start, int count) {
    json questions = json::array();
    for (int index = start; index < start + count; ++index) {
        questions.push_back({{"frontendQuestionId", std::to_string(index)},
                             {"title", "Problem " + std::to_string(index)},
                             {"titleSlug", "problem-" + std::to_string(index)},
                             {"difficulty", "Easy"},
                             {"paidOnly", false}});
    }
    return {{"data",
             {{"problemsetQuestionList", {{"total", total}, {"questions", questions}}}}}};
}

http::TransportFactory factory_for(testing::FakeTransport::Handler handler) {
    return [handler]() -> std::unique_ptr<http::HttpTransport> {
        return std::make_unique<testing::FakeTransport>(handler);
    };
}

auto no_sleep = [](double) {};

}  // namespace

TEST_CASE("fetch paginates until total") {
    std::vector<int> calls;
    testing::FakeTransport transport([&calls](const http::Request& request) {
        int skip = json::parse(request.body)["variables"]["skip"];
        calls.push_back(skip);
        int start = skip + 1;
        int count = std::min(500, 1200 - skip);
        return json_response(200, page(1200, start, count).dump());
    });
    auto questions = problems::fetch_all(transport, no_sleep);
    CHECK(questions.size() == 1200);
    CHECK(calls == std::vector<int>{0, 500, 1000});
    CHECK(questions[0].id == "1");
}

TEST_CASE("fetch advances by actual page size when the server caps it") {
    std::vector<int> skips;
    testing::FakeTransport transport([&skips](const http::Request& request) {
        int skip = json::parse(request.body)["variables"]["skip"];
        skips.push_back(skip);
        int count = std::max(0, std::min(100, 250 - skip));
        return json_response(200, page(250, skip + 1, count).dump());
    });
    auto questions = problems::fetch_all(transport, no_sleep);
    CHECK(skips == std::vector<int>{0, 100, 200});
    REQUIRE(questions.size() == 250);
    CHECK(questions.front().id == "1");
    CHECK(questions.back().id == "250");
}

TEST_CASE("fetch degrades page size then category") {
    std::vector<std::pair<std::string, int>> attempts;
    testing::FakeTransport transport([&attempts](const http::Request& request) {
        auto variables = json::parse(request.body)["variables"];
        attempts.emplace_back(variables["categorySlug"], variables["limit"]);
        if (variables["limit"] == 500) {
            return json_response(400,
                                 json({{"errors", {{{"message", "limit too high"}}}}}).dump());
        }
        if (variables["categorySlug"] == "all-code-essentials") {
            return json_response(200,
                                 json({{"errors", {{{"message", "bad category"}}}}}).dump());
        }
        return json_response(200, page(2, 1, 2).dump());
    });
    auto questions = problems::fetch_all(transport, no_sleep);
    CHECK(questions.size() == 2);
    bool degraded = false;
    for (const auto& [category, limit] : attempts) {
        if (category.empty() && limit == 100) degraded = true;
    }
    CHECK(degraded);
}

TEST_CASE("fetch raises when everything fails") {
    testing::FakeTransport transport([](const http::Request&) {
        http::Response response;
        response.status = 500;
        response.body = "boom";
        return response;
    });
    CHECK_THROWS_AS(problems::fetch_all(transport, no_sleep), LeetCodeError);
}

TEST_CASE("cache round trip") {
    IsolatedDirs dirs;
    auto questions = testing::questions();
    auto path = problems::save_cache(questions, kNow);
    auto content = problems::load_cache(path);
    REQUIRE(content.fetched_at.has_value());
    CHECK(*content.fetched_at == kNow);
    CHECK(content.questions == questions);

    std::ifstream in(path);
    std::ostringstream buffer;
    buffer << in.rdbuf();
    json record = json::parse(buffer.str());
    CHECK(record["fetched_at"] == "2026-07-27T10:00:00Z");
}

TEST_CASE("load cache missing and corrupt") {
    IsolatedDirs dirs;
    auto missing = problems::load_cache();
    CHECK_FALSE(missing.fetched_at.has_value());
    CHECK(missing.questions.empty());

    auto path = paths::problems_cache_file();
    std::filesystem::create_directories(path.parent_path());
    std::ofstream(path) << "{not json";
    auto corrupt = problems::load_cache(path);
    CHECK_FALSE(corrupt.fetched_at.has_value());
    CHECK(corrupt.questions.empty());
}

TEST_CASE("is stale") {
    CHECK(problems::is_stale(std::nullopt, 7));
    CHECK_FALSE(problems::is_stale(kNow, 7, kNow + 6 * kDay));
    CHECK(problems::is_stale(kNow, 7, kNow + 8 * kDay));
    CHECK(problems::is_stale(kNow, 0, kNow));
}

TEST_CASE("get questions uses fresh cache") {
    IsolatedDirs dirs;
    auto questions = testing::questions();
    problems::save_cache(questions, kNow);
    bool network_used = false;
    auto factory = [&network_used]() -> std::unique_ptr<http::HttpTransport> {
        network_used = true;
        return std::make_unique<testing::FakeTransport>(
            [](const http::Request&) { return json_response(200, "{}"); });
    };
    auto result = problems::get_questions(
        {.ttl_days = 7, .transport_factory = factory, .now = kNow});
    CHECK(result == questions);
    CHECK_FALSE(network_used);
}

TEST_CASE("get questions serves stale cache and refreshes in background") {
    IsolatedDirs dirs;
    auto questions = testing::questions();
    problems::save_cache(questions, kNow - 30 * kDay);
    problems::RefreshHandle refresh;
    auto result = problems::get_questions(
        {.ttl_days = 7,
         .transport_factory = factory_for([](const http::Request&) {
             return json_response(200, page(2, 1, 2).dump());
         }),
         .now = kNow},
        &refresh);
    // The stale cache is returned immediately, not the 2-question fetch result.
    CHECK(result == questions);
    CHECK(refresh.active());
    refresh.join();
    CHECK(problems::load_cache().questions.size() == 2);
}

TEST_CASE("background refresh failure keeps the stale cache") {
    IsolatedDirs dirs;
    auto questions = testing::questions();
    problems::save_cache(questions, kNow - 30 * kDay);
    auto handle = problems::spawn_refresh(
        paths::problems_cache_file(), factory_for([](const http::Request&) -> http::Response {
            throw http::TransportError("offline");
        }));
    handle.join();
    CHECK(problems::load_cache().questions == questions);
}

TEST_CASE("get questions without cache or network") {
    IsolatedDirs dirs;
    auto factory = factory_for(
        [](const http::Request&) -> http::Response { throw http::TransportError("offline"); });
    CHECK_THROWS_MATCHES(
        problems::get_questions({.ttl_days = 7, .transport_factory = factory, .now = kNow}),
        LeetCodeError,
        Catch::Matchers::MessageMatches(
            Catch::Matchers::ContainsSubstring("no cached problem list")));
}

TEST_CASE("get questions first run announces the fetch") {
    IsolatedDirs dirs;
    std::vector<std::string> notes;
    problems::get_questions({.ttl_days = 7,
                             .transport_factory = factory_for([](const http::Request&) {
                                 return json_response(200, page(2, 1, 2).dump());
                             }),
                             .info = [&notes](const std::string& note) { notes.push_back(note); },
                             .now = kNow});
    REQUIRE_FALSE(notes.empty());
    CHECK(notes[0].find("first run") != std::string::npos);
}

TEST_CASE("get questions explicit refresh falls back to cache with a warning") {
    IsolatedDirs dirs;
    auto questions = testing::questions();
    problems::save_cache(questions, kNow);
    std::vector<std::string> warnings;
    auto result = problems::get_questions(
        {.ttl_days = 7,
         .refresh = true,
         .transport_factory = factory_for([](const http::Request&) -> http::Response {
             throw http::TransportError("offline");
         }),
         .warn = [&warnings](const std::string& note) { warnings.push_back(note); },
         .now = kNow});
    CHECK(result == questions);
    REQUIRE_FALSE(warnings.empty());
    CHECK(warnings[0].find("using cache") != std::string::npos);
}

TEST_CASE("get questions refresh writes cache") {
    IsolatedDirs dirs;
    auto result = problems::get_questions(
        {.ttl_days = 7,
         .refresh = true,
         .transport_factory = factory_for([](const http::Request&) {
             return json_response(200, page(2, 1, 2).dump());
         }),
         .now = kNow});
    CHECK(result.size() == 2);
    CHECK(std::filesystem::exists(paths::problems_cache_file()));
}

TEST_CASE("padded id handles non numeric") {
    CHECK(Question{"42", "T", "t", "Easy"}.padded_id() == "0042");
    CHECK(Question{"LCP 01", "T", "t", "Easy"}.padded_id() == "LCP 01");
}

File: tests/test_problems.py
from __future__ import annotations

import json
from datetime import UTC, datetime, timedelta

import httpx
import pytest

from lcpush import problems
from lcpush.errors import LeetCodeError
from lcpush.paths import problems_cache_file
from lcpush.problems import (
    Question,
    fetch_all,
    get_questions,
    is_stale,
    load_cache,
    save_cache,
)

NOW = datetime(2026, 7, 27, 10, 0, tzinfo=UTC)


def _page(total, start, count):
    return {
        "data": {
            "problemsetQuestionList": {
                "total": total,
                "questions": [
                    {
                        "frontendQuestionId": str(index),
                        "title": f"Problem {index}",
                        "titleSlug": f"problem-{index}",
                        "difficulty": "Easy",
                        "paidOnly": False,
                    }
                    for index in range(start, start + count)
                ],
            }
        }
    }


def _client(handler):
    return httpx.Client(transport=httpx.MockTransport(handler))


def test_fetch_paginates_until_total():
    calls = []

    def handler(request: httpx.Request) -> httpx.Response:
        body = json.loads(request.content)
        skip = body["variables"]["skip"]
        calls.append(skip)
        start = skip + 1
        count = min(500, 1200 - skip)
        return httpx.Response(200, json=_page(1200, start, count))

    with _client(handler) as client:
        questions = fetch_all(client, sleep=lambda _seconds: None)

    assert len(questions) == 1200
    assert calls == [0, 500, 1000]
    assert questions[0].id == "1"


def test_fetch_advances_by_actual_page_size_when_the_server_caps_it():
    """LeetCode caps pages at 100 however large `limit` is; stride must follow."""
    skips = []

    def handler(request: httpx.Request) -> httpx.Response:
        skip = json.loads(request.content)["variables"]["skip"]
        skips.append(skip)
        count = max(0, min(100, 250 - skip))  # server-side cap, ignoring limit
        return httpx.Response(200, json=_page(250, skip + 1, count))

    with _client(handler) as client:
        questions = fetch_all(client, sleep=lambda _seconds: None)

    assert skips == [0, 100, 200]
    assert len(questions) == 250
    assert [q.id for q in questions] == [str(n) for n in range(1, 251)]


def test_fetch_degrades_page_size_then_category():
    attempts = []

    def handler(request: httpx.Request) -> httpx.Response:
        body = json.loads(request.content)
        variables = body["variables"]
        attempts.append((variables["categorySlug"], variables["limit"]))
        if variables["limit"] == 500:
            return httpx.Response(400, json={"errors": [{"message": "limit too high"}]})
        if variables["categorySlug"] == "all-code-essentials":
            return httpx.Response(200, json={"errors": [{"message": "bad category"}]})
        return httpx.Response(200, json=_page(2, 1, 2))

    with _client(handler) as client:
        questions = fetch_all(client, sleep=lambda _seconds: None)

    assert len(questions) == 2
    assert ("", 100) in attempts


def test_fetch_raises_when_everything_fails():
    def handler(_request):
        return httpx.Response(500, text="boom")

    with _client(handler) as client:
        with pytest.raises(LeetCodeError):
            fetch_all(client, sleep=lambda _seconds: None)


def test_cache_round_trip(questions):
    path = save_cache(questions, now=NOW)
    stamp, loaded = load_cache(path)
    assert stamp == NOW
    assert loaded == questions
    record = json.loads(path.read_text())
    assert record["fetched_at"] == "2026-07-27T10:00:00Z"


def test_load_cache_missing_and_corrupt():
    assert load_cache() == (None, ())
    path = problems_cache_file()
    path.parent.mkdir(parents=True, exist_ok=True)
    path.write_text("{not json", encoding="utf-8")
    assert load_cache(path) == (None, ())


def test_is_stale():
    assert is_stale(None, 7)
    assert not is_stale(NOW, 7, now=NOW + timedelta(days=6))
    assert is_stale(NOW, 7, now=NOW + timedelta(days=8))
    assert is_stale(NOW, 0, now=NOW)


def test_get_questions_uses_fresh_cache(questions):
    save_cache(questions, now=NOW)

    def factory():  # pragma: no cover - must never be called
        raise AssertionError("network was used despite a fresh cache")

    assert get_questions(ttl_days=7, client_factory=factory, now=NOW) == questions


def test_get_questions_serves_stale_cache_and_refreshes_in_background(
    questions, monkeypatch
):
    """A stale cache must never block startup on the network."""
    save_cache(questions, now=NOW - timedelta(days=30))
    spawned = []
    monkeypatch.setattr(
        problems, "_spawn_refresh", lambda target, factory: spawned.append(target)
    )

    def factory():  # pragma: no cover - must never be called in the foreground
        raise AssertionError("stale cache blocked on a foreground fetch")

    result = get_questions(ttl_days=7, client_factory=factory, now=NOW)
    assert result == questions
    assert spawned == [problems_cache_file()]


def test_background_refresh_rewrites_the_cache(questions):
    save_cache(questions, now=NOW - timedelta(days=30))

    def factory():
        def handler(_request):
            return httpx.Response(200, json=_page(2, 1, 2))

        return _client(handler)

    thread = problems._spawn_refresh(problems_cache_file(), factory)
    thread.join(timeout=5)
    _stamp, refreshed = load_cache()
    assert len(refreshed) == 2


def test_background_refresh_failure_keeps_the_stale_cache(questions):
    save_cache(questions, now=NOW - timedelta(days=30))

    def factory():
        def handler(_request):
            raise httpx.ConnectError("offline")

        return _client(handler)

    thread = problems._spawn_refresh(problems_cache_file(), factory)
    thread.join(timeout=5)
    _stamp, kept = load_cache()
    assert kept == questions


def test_get_questions_without_cache_or_network():
    def factory():
        def handler(_request):
            raise httpx.ConnectError("offline")

        return _client(handler)

    with pytest.raises(LeetCodeError) as excinfo:
        get_questions(ttl_days=7, client_factory=factory, now=NOW)
    assert "no cached problem list" in excinfo.value.message


def test_get_questions_first_run_announces_the_fetch():
    notes = []

    def factory():
        def handler(_request):
            return httpx.Response(200, json=_page(2, 1, 2))

        return _client(handler)

    get_questions(ttl_days=7, client_factory=factory, info=notes.append, now=NOW)
    assert notes and "first run" in notes[0]


def test_get_questions_explicit_refresh_falls_back_to_cache_with_a_warning(questions):
    save_cache(questions, now=NOW)
    warnings = []

    def factory():
        def handler(_request):
            raise httpx.ConnectError("offline")

        return _client(handler)

    result = get_questions(
        ttl_days=7, refresh=True, client_factory=factory, warn=warnings.append, now=NOW
    )
    assert result == questions
    assert warnings and "using cache" in warnings[0]


def test_get_questions_refresh_writes_cache():
    def factory():
        def handler(_request):
            return httpx.Response(200, json=_page(2, 1, 2))

        return _client(handler)

    result = get_questions(ttl_days=7, refresh=True, client_factory=factory, now=NOW)
    assert len(result) == 2
    assert problems_cache_file().exists()


def test_padded_id_handles_non_numeric():
    assert Question("42", "T", "t", "Easy").padded_id == "0042"
    assert Question("LCP 01", "T", "t", "Easy").padded_id == "LCP 01"
